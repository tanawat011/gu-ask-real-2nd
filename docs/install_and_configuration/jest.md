# Jest for typescript

```text
`Jest` is a test runner.
```

## Documentation

- [Jest.io](https://jestjs.io/docs/getting-started)
- [testing-library](https://testing-library.com/docs/)

---

## Install and Configuration

### Install Jest for typescript

```sh
yarn add jest ts-jest
```

### Configuration

#### Set up test script in your [`package.json`](../package.json) file

```diff
...
 "scripts": {
    ...
+    "test": "react-scripts test --watchAll=false",
+    "test:w": "react-scripts test",
+    "test:cv": "yarn test --coverage",
+    "test:cv:msg": "yarn test:cv --verbose",
+    "test:cv:w": "yarn test:w --coverage",
+    "test:cv:msg:w": "yarn test:cv:w --verbose",
+    "test:cv:open": "open coverage/lcov-report/index.html",
    ...
  },
...
```

#### Set up basic configuration in your [`package.json`](../package.json) file

```diff
...
+  "jest": {
+    "coverageThreshold": {
+      "global": {
+        "statements": 40,
+        "branches": 40,
+        "functions": 40,
+        "lines": 40
+      }
+    },
+    "coveragePathIgnorePatterns": [
+      "src/pages/*",
+      "src/types/*",
+      "index.tsx",
+      "index.ts"
+    ]
+  }
...
```
