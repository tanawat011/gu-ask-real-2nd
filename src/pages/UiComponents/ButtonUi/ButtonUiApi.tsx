import tw from 'twin.macro'

import { Title } from 'components/PageUiComponent'
import { TableCommon } from 'components/Table'

import { ButtonApiColumns, ButtonApiData } from './data'

const TwContainer = tw.div`mb-8`

export const ButtonUiApi = () => {
  return (
    <TwContainer>
      <Title title='API' />

      <TableCommon columns={ButtonApiColumns} data={ButtonApiData} />
    </TwContainer>
  )
}
