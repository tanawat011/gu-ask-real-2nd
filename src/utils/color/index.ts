export * from './hex2Rgb'
export * from './rgb2Hex'
