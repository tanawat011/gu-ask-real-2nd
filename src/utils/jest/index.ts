export * from './renderWithProviders'
export * from './twColor'
