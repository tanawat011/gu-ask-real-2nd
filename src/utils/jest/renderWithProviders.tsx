import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
