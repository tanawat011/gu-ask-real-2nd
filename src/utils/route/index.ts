export * from './menuGenerator'
export * from './renderRoutes'
export * from './routeGenerator'
