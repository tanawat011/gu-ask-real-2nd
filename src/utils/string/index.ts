export * from './camelToKebab'
export * from './camelToSentence'
export * from './replaceAll'
