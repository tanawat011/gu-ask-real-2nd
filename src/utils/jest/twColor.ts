import { hex2Rgb } from 'utils/color'

export const twColor = (hexColor: string, type?: 'bg' | 'border' | 'text') => {
  if (hexColor === 'transparent') {
    return 'transparent'
  }

  if (hexColor.length > 7) {
    return 'undefined'
  }

  const rgb = hex2Rgb(hexColor, { isArray: true }) as string[]

  if (!type) {
    type = 'bg'
  }

  return `rgb(${rgb.join(' ')} / var(--tw-${type}-opacity))`
}
