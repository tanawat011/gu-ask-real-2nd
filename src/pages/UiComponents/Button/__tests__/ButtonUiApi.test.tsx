import { renderWithProviders } from 'utils/jest'

import { ButtonUiApi } from '../ButtonUiApi'

jest.mock('components/PageUiComponent', () => ({
  Title: () => <div>Title</div>,
}))
jest.mock('components/Table', () => ({
  Table: () => (
    <table>
      <thead>
        <tr>
          <th>1</th>
          <th>2</th>
          <th>3</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>x</td>
          <td>y</td>
          <td>z</td>
        </tr>
      </tbody>
    </table>
  ),
}))

describe('<ButtonUiApi />', () => {
  test('renders correctly', () => {
    expect.assertions(1)

    const { asFragment } = renderWithProviders(<ButtonUiApi />)

    expect(asFragment()).toMatchSnapshot()
  })
})
