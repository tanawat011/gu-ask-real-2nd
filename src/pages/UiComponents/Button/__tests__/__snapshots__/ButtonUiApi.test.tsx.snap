// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonUiApi /> renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  margin-bottom: 2rem;
}

<div
    class="c0"
  >
    <div>
      Title
    </div>
    <table>
      <thead>
        <tr>
          <th>
            1
          </th>
          <th>
            2
          </th>
          <th>
            3
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            x
          </td>
          <td>
            y
          </td>
          <td>
            z
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
